bpo-46613: Add PyType_GetModuleByDef to the public & limited API (#31081)
python/cpython: Doc/c-api/type.rst

   .. versionadded:: 3.9

> .. c:function:: void* PyType_GetModuleByDef(PyTypeObject *type, struct PyModuleDef *def)

Reviewer: The return type is `` PyObject*``, no?

```suggestion
.. c:function::  PyObject* PyType_GetModuleByDef(PyTypeObject *type, struct PyModuleDef *def)
```

   Find the first superclass whose module was created from

>    the given PyModuleDef, and return that module.

Reviewer: ```suggestion
   the given :ctype:`PyModuleDef` *def*, and return that module.
```

Author: Sphinx puts a link in the signature, but why not repeat it.

>    If no module is found, sets :py:class:`TypeError` and returns ``NULL``.

Reviewer: I like "raises" to "sets" is correct, it's just a suggestion:

```suggestion
   If no module is found, raises a :py:class:`TypeError` and returns ``NULL``.
```

   This function is intended to be used together with
   :c:func:`PyModule_GetState()` to get module state from slot methods (such as
   :c:member:`~PyTypeObject.tp_init` or :c:member:`~PyNumberMethods.nb_add`)
   and other places where a method's defining class cannot be passed using the

>    :c:type:`PyCMethod` calling convention.

Reviewer: You may mention somewhere that this function only works on a type created by ``PyType_FromModuleAndSpec``.

Author: It works with any method that attaches the module to the class. Currently there's just one, but I've been adding new functions with extra arguments relatively fast. I'd rather not document one specific function.
